modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -226,26 +232,58 @@ public CompletableFuture<QueryPlan> prepareAsync(
             return CompletableFuture.failedFuture(new SchemaNotFoundException(schemaName));
         }
 
+        long actualPlannerTimeout;
+        Instant deadline = operationContext.operationDeadline();
+        if (deadline == null) {
+            actualPlannerTimeout = plannerTimeout;
+        } else {
+            Instant now = Instant.now();
+            long timeout = Duration.between(now, deadline).toMillis();
+
+            actualPlannerTimeout = Math.min(timeout, plannerTimeout);
+        }
+
         PlanningContext planningContext = PlanningContext.builder()
                 .frameworkConfig(Frameworks.newConfigBuilder(FRAMEWORK_CONFIG).defaultSchema(schema).build())
                 .query(parsedResult.originalQuery())
-                .plannerTimeout(plannerTimeout)
+                .plannerTimeout(actualPlannerTimeout)
                 .parameters(Commons.arrayToMap(operationContext.parameters()))
                 .build();
 
+        // Add an action to trigger planner timeout, when operation times out.
+        // Or trigger timeout immediately if operation has already timed out. 
+        QueryCancel cancelHandler = operationContext.cancel();
+        assert cancelHandler != null;
+
+        Cancellable callback = (timeout) -> {
+            if (timeout) {
+                planningContext.abortByTimeout();
+            }
+        };
+        cancelHandler.add(callback);
```

Comment:
Honestly I don't quite understand why we need this.
Do we have a test that will fail without this change?

Reply:
Please elaborate. I think I explained it in the PR description.

> QueryCancel::add acts as a guard for an operation, because it throws QueryCancelledException, when a query has already been cancelled. E.g. If parsing took too long, then QueryCancel::add in PrepareService throws a QueryCancelledException exception, and the operation terminates. The same happens in ExecutionService.